File: robot-bounded-in-circle.cpp
//Time - O(n)
//Space - O(1)
class Solution {
public:
    bool isRobotBounded(string instructions) {
        vector<vector<int>> dirs {{0,1},{-1,0},{0,-1},{1,0}};
        int dir = 0;
        int x = 0,y = 0;
        for(auto in:instructions){
            if(in == 'G'){
                x = x+dirs[dir][0];
                y = y+dirs[dir][1];
            }else if(in == 'L'){
                dir = (dir+1)%4;
            }else {
                dir = (dir+3)%4;
            }

        }
        
        return (dir!=0 || x == 0 && y==0);
        
    }
};

File: Robot_Bounded_in_Circle.py
# Time Complexity : O(n).
# Space Complexity : O(1);
# Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No
#
#
class Solution:
    def isRobotBounded(self, instructions: str) -> bool:
        dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
        ptr = 0
        position = [0, 0]
        direction = [0, 1]
        for i in instructions:
            if i == 'G':
                position[0] += dirs[ptr][0]
                position[1] += dirs[ptr][1]
            elif i == 'L':
                ptr = (ptr+3) % 4
                direction = dirs[ptr]
            elif i == 'R':
                ptr = (ptr+1) % 4
                direction = dirs[ptr]
        if position == [0, 0]:
            return True
        elif direction != [0, 1]:
            return True
        return False


print(Solution().isRobotBounded("GGLLGG"))


# class Solution:
#     def isRobotBounded(self, instructions: str) -> bool:
#         dirs = [[0, 1], [1, 0], [0, -1], [-1, 0]]
#         ptr = 0
#         position = [0, 0]
#         direction = [0, 1]
#         for _ in range(4):
#             for i in instructions:
#                 if i == 'G':
#                     position[0] += dirs[ptr][0]
#                     position[1] += dirs[ptr][1]
#                 elif i == 'L':
#                     ptr = (ptr+3) % 4
#                     direction = dirs[ptr]
#                 elif i == 'R':
#                     ptr = (ptr+1) % 4
#                     direction = dirs[ptr]
#             if position == [0, 0]:
#                 return True
#         return False
#
#
# print(Solution().isRobotBounded("GGLLGG"))
